### DISCUSSION TITLE
Auto-update translations

### DISCUSSION THREAD

**Discussion by andrey-helldar** [Laravel-Lang/common#48] [Q&A] [answered]
We now have several problems:

1. Many localizations that no one translates (https://github.com/Laravel-Lang/lang/discussions/1389);
2. When creating new keys, you need to look for them in the "en" or "todo.md" file, copy-paste and translate.

I have an idea: what if I create functionality to track such changes and automatically make corrections?

For instance:

- added a new key to the file - the script will automatically insert this key into all localization files.
- the key is not translated in the localization file? - the script will send a request to Google Translate and save the translated value.

Yes, Google Translate doesn't always translate well, but it's better than nothing.

There are two problems we will face:

1. Comments will be removed from the files;
2. Google Translate only allows you to translate 500k characters per month for free. Roughly speaking, the current English language files contain 28,170 characters. In 76 localizations, this will be approximately 2'140'920 characters. Yes, only non-translated values will be sent for translation, but judging by the [todo.md](https://github.com/Laravel-Lang/lang/blob/master/todo.md) file, we have fully translated 3-7 localizations.

I have an idea how to do it quickly and easily.

@Laravel-Lang/laravel-lang, what do you think?

**Comment by andrey-helldar** [comment]
@Laravel-Lang/laravel-lang, guys?

**Comment by mohamedsabil83** [comment, marked as answer]
I talked about that before. I agree with auto insert the new keys to the localization files but without translations. It's enough to find the new keys in the localization files and translate them from English.

Also, we can use google translation for localizations that didn't have a translator yet.

**Reply by andrey-helldar** [reply]
Several people have talked about this idea. I can implement it one of these days.
There is another point about the relevance of the keys - if you delete a key from the English localization, it will automatically be deleted from all localizations. Do we need this?

**Reply by mohamedsabil83** [reply]
IMHO, I think so because the English file is the reference for all keys.

In my last commit, I found a line in the file **ar.json** and not in the **en.json**. I returned it to the **en.json** as I thought someone delete it accidentally but one of the contributors notice me that the key did not exist anymore in the Laravel core file.

So, I think it's ok to auto delete key(s) from all translation files when deleted from the English files.

**Reply by caouecs** [reply]
A lot of tests to accept the script ;)

but I agree

**Reply by andrey-helldar** [reply]
@mohamedsabil83, I also came across this when I compared `ru` localization with `en`.

Okay, then I'll make the necessary changes one of these days.

**Reply by luisprmat** [reply]
In this moment I find it convenient to have an automated test suite with PHPUnit for `todo.php` script for testing the new features without damaging the previous ones and to be able to refactor with greater confidence.

**Reply by andrey-helldar** [reply]
Done! https://github.com/Laravel-Lang/lang/pull/1560

**Reply by andrey-helldar** [reply]
@luisprmat, testing a translation is bad, because we know in advance that not all keys have been translated.

You can test the todo.md file for the `## <lang>` line, but why? ..

The only thing I see good for testing is the key match in the files.

**Reply by andrey-helldar** [reply]
@luisprmat, i added tests: https://github.com/Laravel-Lang/lang/pull/1560
